src/bloqade/analysis/fidelity/impls.py

```diff
@@ -0,0 +1,100 @@
+        current_gate_fidelity = interp._current_gate_fidelity
+        current_atom_survival = interp._current_atom_survival_probability
+
+        for s in stmt.then_body.stmts():
+            stmt_impl = interp.lookup_registry(frame=frame, stmt=s)
+            if stmt_impl is None:
+                continue
+
+            stmt_impl(interp=interp, frame=frame, stmt=s)
+
+        then_gate_fidelity = interp._current_gate_fidelity
+        then_atom_survival = interp._current_atom_survival_probability
+
+        # NOTE: reset fidelity of interp to check if the else body results in a worse fidelity
+        interp._current_gate_fidelity = current_gate_fidelity
+        interp._current_atom_survival_probability = current_atom_survival
+
+        for s in stmt.else_body.stmts():
+            stmt_impl = interp.lookup_registry(frame=frame, stmt=s)
+            if stmt_impl is None:
+                continue
+
+            stmt_impl(interp=interp, frame=frame, stmt=s)
+
+        else_gate_fidelity = interp._current_gate_fidelity
+        else_atom_survival = interp._current_atom_survival_probability
+
+        # NOTE: look for the "worse" branch
+        then_combined_fidelity = then_gate_fidelity * then_atom_survival
+        else_combined_fidelity = else_gate_fidelity * else_atom_survival
+
+        if then_combined_fidelity < else_combined_fidelity:
+            interp._current_gate_fidelity = then_gate_fidelity
+            interp._current_atom_survival_probability = then_atom_survival
+        else:
+            interp._current_gate_fidelity = else_gate_fidelity
+            interp._current_atom_survival_probability = else_atom_survival
+
+    @interp.impl(Yield)
+    def yield_(
+        self, interp: FidelityAnalysis, frame: interp.Frame[EmptyLattice], stmt: Yield
+    ):
+        # NOTE: yield can by definition only contain values, never any stmts, so fidelity cannot decrease
+        return
+
+    @interp.impl(For)
+    def for_loop(
+        self, interp: FidelityAnalysis, frame: interp.Frame[EmptyLattice], stmt: For
+    ):
+
+        if not isinstance(hint := stmt.iterable.hints.get("const"), const.Value):
+            # NOTE: not clear how long this loop is
+            # TODO: should we at least count the body once?
+            return
+
+        current_gate_fidelity = interp._current_gate_fidelity
+        current_atom_survival = interp._current_atom_survival_probability
+
+        # NOTE: we reset the interpreter fidelity and evaluate the fidelity for the body only once
+        interp._current_gate_fidelity = 1.0
+        interp._current_atom_survival_probability = 1.0
+        for s in stmt.body.stmts():
+            stmt_impl = interp.lookup_registry(frame=frame, stmt=s)
+            if stmt_impl is None:
+                continue
+
+            stmt_impl(interp=interp, frame=frame, stmt=s)
+
+        loop_body_gate_fidelity = interp._current_gate_fidelity
+        loop_body_atom_survival = interp._current_atom_survival_probability
```

Comment:
logically this makes sense but I think you're reinventing the wheel in terms of executing the loop body.Take a look at how kirin implements the interpreters for the structural control flow: https://github.com/QuEraComputing/kirin/blob/2bd47e78eee36a3c703a01655e1af0f5e95492e0/src/kirin/dialects/scf/interp.py#L24
This can also be used for the abstract interpreter.